flink/v1.15/flink/src/main/java/org/apache/iceberg/flink/sink/v2/FilesCommitter.java

```diff
@@ -0,0 +1,224 @@
+import org.apache.flink.api.connector.sink2.Committer;
+import org.apache.iceberg.AppendFiles;
+import org.apache.iceberg.ReplacePartitions;
+import org.apache.iceberg.RowDelta;
+import org.apache.iceberg.SnapshotUpdate;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.flink.TableLoader;
+import org.apache.iceberg.io.WriteResult;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.util.ThreadPools;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+class FilesCommitter implements Committer<FilesCommittable>, Serializable {
+  private static final long serialVersionUID = 1L;
+
+  private static final Logger LOG = LoggerFactory.getLogger(FilesCommitter.class);
+  private static final String FLINK_JOB_ID = "flink.job-id";
+
+  // TableLoader to load iceberg table lazily.
+  private final TableLoader tableLoader;
+  private final boolean replacePartitions;
+  private final Map<String, String> snapshotProperties;
+
+  // It will have an unique identifier for one job.
+  private final Table table;
+  private transient int continuousEmptyCheckpoints;
+
+  private transient ExecutorService workerPool;
+
+  FilesCommitter(TableLoader tableLoader, boolean replacePartitions, Map<String, String> snapshotProperties,
+                 int workerPoolSize) {
+    this.tableLoader = tableLoader;
+    this.replacePartitions = replacePartitions;
+    this.snapshotProperties = snapshotProperties;
+
+    this.workerPool =
+        ThreadPools.newWorkerPool("iceberg-worker-pool-" + Thread.currentThread().getName(), workerPoolSize);
+
+    // Open the table loader and load the table.
+    this.tableLoader.open();
+    this.table = tableLoader.loadTable();
+  }
+
+  @Override
+  public void commit(Collection<CommitRequest<FilesCommittable>> requests)
+      throws IOException, InterruptedException {
+    int dataFilesNumRestored = 0;
+    int deleteFilesNumRestored = 0;
+
+    int dataFilesNumStored = 0;
+    int deleteFilesNumStored = 0;
+
+    Collection<CommitRequest<FilesCommittable>> restored = Lists.newArrayList();
+    Collection<CommitRequest<FilesCommittable>> store = Lists.newArrayList();
+
+    for (CommitRequest<FilesCommittable> request : requests) {
+      WriteResult committable = request.getCommittable().committable();
+      if (request.getNumberOfRetries() > 0) {
+        restored.add(request);
+        dataFilesNumRestored = dataFilesNumRestored + committable.dataFiles().length;
+        deleteFilesNumRestored = deleteFilesNumRestored + committable.deleteFiles().length;
+      } else {
+        store.add(request);
+        dataFilesNumStored = dataFilesNumStored + committable.dataFiles().length;
+        deleteFilesNumStored = deleteFilesNumStored + committable.deleteFiles().length;
+      }
+    }
+
+    if (restored.size() > 0) {
+      commitResult(dataFilesNumRestored, deleteFilesNumRestored, restored);
```

Comment:
I am not sure what happens in the Iceberg code, if we want to add already existing files to the table.

Based on the Flink doc, the commit should be idempotent, so if the previous Iceberg commit failed with `CommitStateUnknownException` because there were some communication error with the Catalog then we should be able to commit the changes again (even if the commit was successful previously, or even if it was unsuccessful).

Do we have test for this case?

Reply:
Sorry for the late reply. I've been thinking about what you said.
The goal should be to avoid writing duplicate data. So our best bet was to come up with a mechanism that would allow us to submit the data exactly one way.

I will test the failed commit case again soon.

Thank you.